pkg/solver/che_routing.go

```diff
@@ -0,0 +1,474 @@
+	configMaps, err := c.getGatewayConfigsAndFillRoutingObjects(cheManager, workspaceMeta.WorkspaceId, routing, &objs)
+	if err != nil {
+		return solvers.RoutingObjects{}, err
+	}
+
+	syncer := sync.New(c.client, c.scheme)
+
+	for _, cm := range configMaps {
+		_, _, err := syncer.Sync(context.TODO(), nil, &cm, configMapDiffOpts)
+		if err != nil {
+			return solvers.RoutingObjects{}, err
+		}
+
+	}
+
+	return objs, nil
+}
+
+func (c *CheRoutingSolver) cheExposedEndpoints(manager *dwoche.CheManager, workspaceID string, endpoints map[string]dwo.EndpointList, routingObj solvers.RoutingObjects) (exposedEndpoints map[string]dwo.ExposedEndpointList, ready bool, err error) {
+	if manager.Status.GatewayPhase != dwoche.GatewayPhaseEstablished {
+		return nil, false, nil
+	}
+
+	gatewayHost := manager.Status.GatewayHost
+
+	exposed := map[string]dwo.ExposedEndpointList{}
+
+	for machineName, endpoints := range endpoints {
+		exposedEndpoints := dwo.ExposedEndpointList{}
+		for _, endpoint := range endpoints {
+			if endpoint.Exposure != dw.PublicEndpointExposure {
+				continue
+			}
+
+			var scheme string
+			if endpoint.Protocol == "" {
+				scheme = "http"
+			} else {
+				scheme = string(endpoint.Protocol)
+			}
+
+			if !isExposableScheme(scheme) {
+				// we cannot expose non-http endpoints publicly, because ingresses/routes only support http(s)
+				continue
+			}
+
+			if endpoint.Secure {
+				scheme = secureScheme(scheme)
+
+				// TODO this should also do the magic of ensuring user authentication however we are going to do it
+				// in the future
+			}
+
+			// try to find the endpoint in the ingresses/routes first. If it is there, it is exposed on a subdomain
+			// otherwise it is exposed through the gateway
+			var endpointURL string
+			if infrastructure.IsOpenShift() {
+				route := findRouteForEndpoint(machineName, endpoint, &routingObj)
+				if route != nil {
+					endpointURL = path.Join(route.Spec.Host, endpoint.Path)
+				}
+			} else {
+				ingress := findIngressForEndpoint(machineName, endpoint, &routingObj)
+				if ingress != nil {
+					endpointURL = path.Join(ingress.Spec.Rules[0].Host, endpoint.Path)
+				}
+			}
+
+			if endpointURL == "" {
+				if manager.Spec.GatewayDisabled {
+					return map[string]dwo.ExposedEndpointList{}, false, fmt.Errorf("couldn't find an ingress/route for an endpoint `%s` in workspace `%s`, this is a bug", endpoint.Name, workspaceID)
+				}
+
+				if gatewayHost == "" {
+					// the gateway has not yet established the host
+					return map[string]dwo.ExposedEndpointList{}, false, nil
+				}
+
+				publicURLPrefix := getPublicURLPrefixForEndpoint(workspaceID, machineName, endpoint)
+				endpointURL = path.Join(gatewayHost, publicURLPrefix, endpoint.Path)
+			}
+
+			publicURL := scheme + "://" + endpointURL
+
+			// path.Join() removes the trailing slashes, so make sure to reintroduce that if required.
+			if endpoint.Path == "" || strings.HasSuffix(endpoint.Path, "/") {
```

Comment:
shouldn't second condition be negated? If path does NOT have "/" ?

Reply:
No. If the `endpoint.Path` is empty, we assume it defaults to `/`. We therefore need to add the `/` back, because `path.Join()` removed it. This also needs to be the case if the `endpoint.Path` ends with a `/`. When composing the full URL, `path.Join()` removed the trailing `/` but because the `endpoint.Path` contains it, we add it back.

Reply:
oh I see it now